public/app/plugins/datasource/testdata/datasource.ts

```diff
@@ -34,9 +38,57 @@ export class TestDataDataSource extends DataSourceApi<TestDataQuery> {
     private readonly templateSrv: TemplateSrv = getTemplateSrv()
   ) {
     super(instanceSettings);
+
+    this.variables = {
+      toMetricFindValues: () => dataStream =>
+        dataStream.pipe(
+          mergeMap(data => {
+            if (!data || data.length === 0) {
+              return of([]);
+            }
+
+            const frame = data[0];
+            const values: MetricFindValue[] = [];
+            for (let index = 0; index < frame.length; index++) {
+              const text = frame.fields[0].values.get(index);
+              const value = frame.fields[1].values.get(index);
+              values.push({ text, value });
+            }
+
+            return of(values);
+          })
+        ),
+      toDataQueryRequest: (query, scopedVars) => {
+        const targets = [
+          {
+            datasource: instanceSettings.name,
+            refId: `variable-query-${instanceSettings.name}`,
+            query,
+          },
+        ];
+
+        const request: DataQueryRequest = {
+          targets,
+          app: CoreApp.Variables,
+          range: DefaultTimeRange,
+          scopedVars,
+          requestId: uuidv4(),
+          intervalMs: 0,
+          timezone: 'utc',
+          interval: '',
+          startTime: Date.now(),
+        };
+
+        return request;
+      },
+    };
   }
 
   query(options: DataQueryRequest<TestDataQuery>): Observable<DataQueryResponse> {
+    if (options.app === CoreApp.Variables) {
+      return this.handleVariablesQuery(options);
+    }
+
```

Comment:
I'm still not super convinced that this is a good API pattern. I think having a specific query function for variables is better. This is just the changes in TestData. I think I'll try to expand this PR to include prometheus as well and see how good/bad that would be.

Reply:
If we can avoid the query ever checking `options.app` that would make me happier.  I see this is just in the testdata implementation, so I'm not sure how fundamental/required the app check is.

The use of options.app changing behavior causes a lot of inconsistency that is hard to track down.

Reply:
Great catch! Maybe we don't need to use this anymore when I think about it. I introduced the toDataQueryRequest to the variable support API so that should be enough. Thank you Ryan!